public/apps/configuration/panels/role-edit.tsx

```diff
@@ -47,24 +62,289 @@ const TITLE_TEXT_DICT = {
+  onFieldChangeHandler: (s: OptionSeletion) => void;
+  onFieldCreateHandler: (s: string) => void;
+}) {
+  return (
+    <FormRow
+      headerText="Field level security"
+      headerSubText="You can restrict what document field that user can see. If you use field level security in conjunction with document-level security, make sure you don't restrict to the field that document-level security uses."
+      optional
+    >
+      <EuiFlexGroup>
+        <EuiFlexItem grow={1}>
+          <EuiSuperSelect
+            valueOfSelected={props.method}
+            options={[
+              { inputDisplay: 'Include', value: 'include' },
+              { inputDisplay: 'Exclude', value: 'exclude' },
+            ]}
+            onChange={props.onMethodChangeHandler}
+          />
+        </EuiFlexItem>
+        <EuiFlexItem grow={9}>
+          <EuiComboBox
+            noSuggestions
+            placeholder="Type in field name"
+            selectedOptions={props.fields}
+            onChange={props.onFieldChangeHandler}
+            onCreateOption={props.onFieldCreateHandler}
+          />
+        </EuiFlexItem>
+      </EuiFlexGroup>
+    </FormRow>
+  );
+}
+
+function AnonymizationRow(props: {
+  value: OptionSeletion;
+  onChangeHandler: (s: OptionSeletion) => void;
+  onCreateHandler: (s: string) => void;
+}) {
+  return (
+    <FormRow
+      headerText="Anonymization"
+      headerSubText="Masks any sensitive fields with random value to protect your data security."
+      optional
+    >
+      <EuiComboBox
+        noSuggestions
+        placeholder="Type in field name"
+        selectedOptions={props.value}
+        onChange={props.onChangeHandler}
+        onCreateOption={props.onCreateHandler}
+      />
+    </FormRow>
+  );
+}
+
+function generateIndexPermissionPanels(
+  indexPermissions: RoleIndexPermissionStateClass[],
+  permisionOptionsSet: EuiComboBoxOptionOption[],
+  setRoleIndexPermission: Dispatch<SetStateAction<RoleIndexPermissionStateClass[]>>
+) {
+  const panels = indexPermissions.map((perm, index) => {
+    const onValueChangeHandler = (field: string) =>
+      updateElementInArrayHandler(setRoleIndexPermission, [index, field]);
+
+    const onCreateOptionHandler = (field: string) =>
+      appendOptionToComboBoxHandler(setRoleIndexPermission, [index, field]);
+
+    return (
+      <Fragment key={`index-permission-${index}`}>
+        <EuiAccordion
+          id={`index-permission-${index}`}
+          buttonContent={
+            perm.indexPatterns.map(comboBoxOptionToString).join(', ') || 'Add index permission'
+          }
+          extraAction={
+            <EuiButton
+              color="danger"
+              onClick={() => removeElementFromArray(setRoleIndexPermission, [], index)}
+            >
+              Remove
+            </EuiButton>
+          }
+        >
+          <IndexPatternRow
+            value={perm.indexPatterns}
+            onChangeHandler={onValueChangeHandler('indexPatterns')}
```

Comment:
For all the attribute names mentioned in this file, if there is any typo, it will lead to a UI bug. I would suggest to create some constant strings to have a centralized place to management instead of hardcoding them around.

Reply:
I don't think that is necessary, currently every field is referred in strings for only twice (in change and create handler) and it is centralized in this function.